go/database/mpt/nodes_test.go

```diff
@@ -6554,6 +6554,326 @@ func markModifiedAsDirty(t *testing.T, ctxt *nodeContext, before, after NodeRefe
 	handle.Release()
 }
 
+func TestIterateAccountNodes_CanIterateNodesByCorrectAddresses(t *testing.T) {
+	var address common.Address
+	for i := range address {
+		address[i] = byte(i)
+	}
+
+	path := addressToNibbles(address)
+	rootId := NewNodeReference(NodeId(500))
+	accountId := NewNodeReference(AccountId(100))
+	account := &AccountNode{address: address}
+
+	tests := []struct {
+		name string
+		mock func(source *MockNodeSource, visitor *MockNodeVisitor)
+	}{{
+		"B-E-B",
+		func(source *MockNodeSource, visitor *MockNodeVisitor) {
```

Comment:
By reusing the available test infrastructure, I think those tests could be made way more readable. Something like this:
```
func TestIteratorAccountNodes_Tryout(t *testing.T) {
	address := common.Address{0x12, 0x34, 0x56, 0x78}

	tests := map[string]struct {
		trie    NodeDesc
		path    []string
		present bool
	}{
		"empty": {
			trie: &Tag{"A", &Empty{}},
			path: []string{"A"},
		},
		"just a wrong account": {
			trie: &Tag{"A", &Account{}},
			path: []string{"A"}, // < this should be empty
		},
		"just the correct account": {
			trie:    &Tag{"A", &Account{address: address}},
			path:    []string{"A"},
			present: true,
		},
		"branch without account": {
			trie: &Tag{"A", &Branch{children: Children{
				0: &Tag{"B", &Empty{}},
				1: &Tag{"C", &Empty{}},
				2: &Tag{"D", &Empty{}},
			}}},
			path: []string{"A", "C"}, // < this should just be "A"
		},
		"branch with wrong account": {
			trie: &Tag{"A", &Branch{children: Children{
				0: &Tag{"B", &Empty{}},
				1: &Tag{"C", &Account{}},
				2: &Tag{"D", &Empty{}},
			}}},
			path: []string{"A", "C"}, // < this should just be "A"
		},
		"branch with correct account": {
			trie: &Tag{"A", &Branch{children: Children{
				0: &Tag{"B", &Empty{}},
				1: &Tag{"C", &Account{address: address}},
				2: &Tag{"D", &Empty{}},
			}}},
			path:    []string{"A", "C"},
			present: true,
		},
		"extension with common prefix": {
			trie: &Tag{"A", &Extension{
				path: []Nibble{1, 2, 3},
				next: &Tag{"B", &Branch{children: Children{
					3: &Tag{"C", &Empty{}},
					4: &Tag{"D", &Empty{}},
				}},
				}}},
			path: []string{"A", "B", "C"}, // < this should just be "A" and "B"
		},
		"extension without common prefix": {
			trie: &Tag{"A", &Extension{path: []Nibble{2, 3}}},
			path: []string{"A"}, // < this should just be nothing
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			ctrl := gomock.NewController(t)
			ctxt := newNiceNodeContext(t, ctrl)

			root, _ := ctxt.Build(test.trie)

			visitor := NewMockNodeVisitor(ctrl)
			var last *gomock.Call
			for _, label := range test.path {
				ref, shared := ctxt.Get(label)
				handle := shared.GetReadHandle()
				cur := visitor.EXPECT().Visit(handle.Get(), NodeInfo{Id: ref.Id()})
				handle.Release()
				if last != nil {
					cur.After(last)
				}
				last = cur
			}

			found, err := IterateAccountNodes(ctxt, &root, address, visitor)
			if err != nil {
				t.Fatalf("unexpected error during path iteration: %v", err)
			}
			if found != test.present {
				t.Errorf("unexpected found result, wanted %t, got %t", test.present, found)
			}
		})
	}
}
```
What do you think?

Reply:
A slight improvement, automating the detection of the presence of the test account:
```
func TestIteratorAccountNodes_Tryout(t *testing.T) {
	address := common.Address{0x12, 0x34, 0x56, 0x78}

	tests := map[string]struct {
		trie NodeDesc // < the structure of the trie
		path []string // < the path to follow to reach the test account
	}{
		"empty": {
			trie: &Tag{"A", &Empty{}},
			path: []string{"A"},
		},
		"just a wrong account": {
			trie: &Tag{"A", &Account{}},
			path: []string{"A"}, // < this should be empty
		},
		"just the correct account": {
			trie: &Tag{"A", &Account{address: address}},
			path: []string{"A"},
		},
		"branch without account": {
			trie: &Tag{"A", &Branch{children: Children{
				0: &Tag{"B", &Empty{}},
				1: &Tag{"C", &Empty{}},
				2: &Tag{"D", &Empty{}},
			}}},
			path: []string{"A", "C"}, // < this should just be "A"
		},
		"branch with wrong account": {
			trie: &Tag{"A", &Branch{children: Children{
				0: &Tag{"B", &Empty{}},
				1: &Tag{"C", &Account{}},
				2: &Tag{"D", &Empty{}},
			}}},
			path: []string{"A", "C"}, // < this should just be "A"
		},
		"branch with correct account": {
			trie: &Tag{"A", &Branch{children: Children{
				0: &Tag{"B", &Empty{}},
				1: &Tag{"C", &Account{address: address}},
				2: &Tag{"D", &Empty{}},
			}}},
			path: []string{"A", "C"},
		},
		"extension with common prefix": {
			trie: &Tag{"A", &Extension{
				path: []Nibble{1, 2, 3},
				next: &Tag{"B", &Branch{children: Children{
					3: &Tag{"C", &Empty{}},
					4: &Tag{"D", &Empty{}},
				}},
				}}},
			path: []string{"A", "B", "C"}, // < this should just be "A" and "B"
		},
		"extension without common prefix": {
			trie: &Tag{"A", &Extension{path: []Nibble{2, 3}}},
			path: []string{"A"}, // < this should just be nothing
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			ctrl := gomock.NewController(t)
			ctxt := newNiceNodeContext(t, ctrl)

			root, shared := ctxt.Build(test.trie)

			visitor := NewMockNodeVisitor(ctrl)
			var last *gomock.Call
			for _, label := range test.path {
				ref, shared := ctxt.Get(label)
				handle := shared.GetReadHandle()
				cur := visitor.EXPECT().Visit(handle.Get(), NodeInfo{Id: ref.Id()})
				handle.Release()
				if last != nil {
					cur.After(last)
				}
				last = cur
			}

			found, err := IterateAccountNodes(ctxt, &root, address, visitor)
			if err != nil {
				t.Fatalf("unexpected error during path iteration: %v", err)
			}

			accountPresent := false
			handle := shared.GetReadHandle()
			handle.Get().Visit(ctxt, &root, 0, MakeVisitor(func(n Node, i NodeInfo) VisitResponse {
				if node, ok := n.(*AccountNode); ok && node.address == address {
					accountPresent = true
				}
				return VisitResponseContinue
			}))
			handle.Release()

			if found != accountPresent {
				t.Errorf("unexpected found result, wanted %t, got %t", accountPresent, found)
			}
		})
	}
}
```

Reply:
much better! used this